File: 修订补充示例/第二轮/ch08_正文代码框.R
# 正文代码框顺序运行；扩展数据凡带“模拟”均非真实调查记录。
args <- commandArgs(trailingOnly=FALSE)
file_arg <- args[grepl("^--file=",args)]
script_dir <- if(length(file_arg)) dirname(normalizePath(gsub("~+~"," ",sub("^--file=","",file_arg[1]),fixed=TRUE))) else getwd()
repo_dir <- normalizePath(file.path(script_dir,"..",".."))
setwd(repo_dir)
library(sandwich)
library(lmtest)
set.seed(922)


# 代码框 1: book_chapters/ch08_dependent_errors.tex:92 / ch08_dependent_errors_0
data_path <- file.path("Chapter08_Regression_analysis_with_dependent_error_terms_ch_reg_dep_error",
                       "results", "chapter08_acs_cluster_analysis_data.csv")
acs <- read.csv(data_path, fileEncoding="UTF-8-BOM")
mod <- lm(log_income ~ bachelor, data = acs)
library(sandwich)
library(lmtest)
V <- vcovCL(mod, cluster = acs$household_id, type = "HC0")
coeftest(mod, vcov. = V, df = length(unique(acs$household_id)) - 1)


# 代码框 2: book_chapters/ch08_dependent_errors.tex:568 / ch08_survey_r
library(survey)
# 模拟数据，仅展示分层与权重接口
df <- read.csv("修订补充示例/第二轮/ch08_调查加权_模拟.csv")

# 设定调查设计对象（df 中：weight 为个体抽样权重，stratum 为分层变量）
# 真实调查须使用技术文件指定的权重、分层和 PSU；地区列不自动等于分层变量
svy_design <- svydesign(ids = ~1,          # 无整群结构时 ~1
                        strata = ~stratum,
                        weights = ~weight,
                        data = df)

# 调查加权回归：标准误已使用分层设计修正
m_svy <- svyglm(income ~ educ + age + urban,
                design = svy_design)
summary(m_svy)   # Estimate 列是 \hat\beta_j；Std. Error 列是分层修正后 SE

# 与不加权 OLS 对比
m_ols <- lm(income ~ educ + age + urban, data = df)
# 比较两列系数与 SE：差异反映权重影响，不能单凭差异判定总体代表性

File: 修订补充示例/第二轮/ch08_正文代码框.py
# 正文代码框顺序运行；带“模拟”的扩展数据不是真实调查记录。
from pathlib import Path
import numpy as np,pandas as pd,statsmodels.formula.api as smf
from scipy import stats
repo_dir=Path(__file__).resolve().parents[2]
import os
os.chdir(repo_dir)
acs=pd.read_csv(repo_dir / 'Chapter08_Regression_analysis_with_dependent_error_terms_ch_reg_dep_error/results/python_chapter08_acs_cluster_analysis_data.csv')
df=pd.read_csv(repo_dir / '修订补充示例/第二轮/ch08_调查加权_模拟.csv')

# 代码框 1: {Python 中按家庭聚类标准误}
import numpy as np, pandas as pd
import statsmodels.formula.api as smf
from pathlib import Path
data_path = (Path("Chapter08_Regression_analysis_with_dependent_error_terms_ch_reg_dep_error")
             / "results" / "chapter08_acs_cluster_analysis_data.csv")
acs = pd.read_csv(data_path)
mod = smf.ols("log_income ~ bachelor", data=acs).fit()
clustered = mod.get_robustcov_results(
    cov_type="cluster", groups=acs["household_id"],
    use_correction=False, df_correction=True, use_t=True
)
G = acs["household_id"].nunique()
# R 的 HC0 + cadjust=TRUE：只乘 G/(G-1)
clustered.cov_params_default *= G / (G - 1)
clustered.summary()


# 代码框 2: [label=ch08_survey_py]{Python 中调查加权回归（示意代码）}
import statsmodels.formula.api as smf
import numpy as np

# 模拟数据，仅展示独立观测下 WLS+HC1
df = pd.read_csv("修订补充示例/第二轮/ch08_调查加权_模拟.csv")
# 方法一：WLS（指定权重）；df 列含义同 R 版
m_wls = smf.wls("income ~ educ + age + urban",
                data=df, weights=df["weight"]).fit(
                cov_type="HC1")
print(m_wls.summary().tables[1])   # coef、std err、t、P>|t|、95% CI

# 方法二：用加权统计量做描述性分析（不输出回归系数）
# 复杂调查设计应使用支持分层与整群的专门调查分析工具
from statsmodels.stats.weightstats import DescrStatsW
ds = DescrStatsW(df[["income","educ"]], weights=df["weight"])
print(ds.mean)   # 加权均值（用于描述统计，与 svyglm 的系数解释一致性）

